In [ ]:
import os
import time
import numpy as np
import stim
from pathlib import Path
import pandas as pd

import cudf
# import cupy as cp
import os
import time
import numpy as np
import stim
from pathlib import Path
import pandas as pd

def sample_values4(input_file_path: Path, shots: int):
    """
    Samples values from a given STIM circuit file.
    
    Args:
        input_file_path (Path): Path to the STIM circuit file.
        shots (int): Number of samples to generate.
    
    Returns:
        tuple: Unique sampled values and their counts.
    """
    with open(input_file_path, 'r', encoding='utf-8') as file:
        content = file.read()
    
    circ = stim.Circuit(content)
    values = circ.compile_detector_sampler().sample(shots=shots, append_observables=True)
    print("finish sampling")
    start_time = time.time()
    df = cudf.DataFrame(values)
    unique_counts = df.value_counts().reset_index()
    print(f"Process Time: {time.time() - start_time:.2f} seconds\n")
    
    return unique_counts

def main():
    code_tasks = ["bivariate_bicycle_code:rotated_memory_z"]
    nkds = [[72, 12, 6], [90, 8, 10], [108, 8, 10], [144, 12, 12]]
    # probabilities = range(5, 30, 5)
    noise_models = ["si1000"]
    shots = 10**7
    print("shots:", shots)
    
    for code_task in code_tasks:
        for nkd in nkds:
            # for p in probabilities:
            p = 20
            r = 1
            folder = "Z" if "memory_z" in code_task else "X" if "memory_x" in code_task else "other"
            # if p==10:
            for noise_model in noise_models:
                # p=10，此时为输入。
                print(f"Running for {code_task}, nkd={nkd}, probability={p}, noise_model={noise_model}")
                
                current_file_directory = os.path.abspath("/home/normaluser/ck/epmld/data/external/epmld_experiment_data/paper_experiment_bb_codes")
                print(current_file_directory)
                input_file_path = os.path.join(current_file_directory, folder, f"nkd_{nkd[0]}_{nkd[1]}_{nkd[2]}_r{r}", f"circuit_noisy_{noise_model}_p{p}.stim")
                
                start_time = time.time()
                unique_counts = sample_values4(input_file_path, shots=shots)
                print("unique_counts 2:", unique_counts.shape)
                print(f"Execution Time: {time.time() - start_time:.2f} seconds\n")

In [2]:
main()

shots: 10000000
Running for bivariate_bicycle_code:rotated_memory_z, nkd=[72, 12, 6], probability=20, noise_model=si1000
/home/normaluser/ck/epmld/data/external/epmld_experiment_data/paper_experiment_bb_codes
finish sampling
Process Time: 4.07 seconds

unique_counts 2: (3692288, 85)
Execution Time: 12.28 seconds

Running for bivariate_bicycle_code:rotated_memory_z, nkd=[90, 8, 10], probability=20, noise_model=si1000
/home/normaluser/ck/epmld/data/external/epmld_experiment_data/paper_experiment_bb_codes
finish sampling
Process Time: 4.92 seconds

unique_counts 2: (5468377, 99)
Execution Time: 14.63 seconds

Running for bivariate_bicycle_code:rotated_memory_z, nkd=[108, 8, 10], probability=20, noise_model=si1000
/home/normaluser/ck/epmld/data/external/epmld_experiment_data/paper_experiment_bb_codes
finish sampling
Process Time: 7.13 seconds

unique_counts 2: (6965358, 117)
Execution Time: 20.18 seconds

Running for bivariate_bicycle_code:rotated_memory_z, nkd=[144, 12, 12], probability=2

shots: 10000000
Running for bivariate_bicycle_code:rotated_memory_z, nkd=[72, 12, 6], probability=20, noise_model=si1000
/home/normaluser/ck/epmld/data/external/epmld_experiment_data/paper_experiment_bb_codes
finish sampling
Process Time: 4.07 seconds

unique_counts 2: (3692288, 85)
Execution Time: 12.28 seconds

Running for bivariate_bicycle_code:rotated_memory_z, nkd=[90, 8, 10], probability=20, noise_model=si1000
/home/normaluser/ck/epmld/data/external/epmld_experiment_data/paper_experiment_bb_codes
finish sampling
Process Time: 4.92 seconds

unique_counts 2: (5468377, 99)
Execution Time: 14.63 seconds

Running for bivariate_bicycle_code:rotated_memory_z, nkd=[108, 8, 10], probability=20, noise_model=si1000
/home/normaluser/ck/epmld/data/external/epmld_experiment_data/paper_experiment_bb_codes
finish sampling
Process Time: 7.13 seconds

unique_counts 2: (6965358, 117)
Execution Time: 20.18 seconds

Running for bivariate_bicycle_code:rotated_memory_z, nkd=[144, 12, 12], probability=20, noise_model=si1000
/home/normaluser/ck/epmld/data/external/epmld_experiment_data/paper_experiment_bb_codes
finish sampling
Process Time: 11.97 seconds

unique_counts 2: (8836556, 157)
Execution Time: 27.49 seconds


在10^7次采样中，我们可以看到，对于不同的nkd，采样的时间是不同的。